examples/gpact/hotel-train/contracts/src/Train.sol

```diff
@@ -0,0 +1,201 @@
+        approvedTravelAgencies[_blockchainId] = _travelAgencyContract;
+        travelAgencySpender[_travelAgencyContract] = spendingAccount;
+    }
+
+    function addSeats(uint256 _seatRate, uint256 _numberOfSeats)
+        external
+        onlyOwner
+    {
+        for (uint256 i = 0; i < _numberOfSeats; i++) {
+            setMapValue(SEAT_RATE_MAP, numSeats++, _seatRate);
+        }
+    }
+
+    function changeSeatRate(uint256 _seatNumber, uint256 _seatRate)
+        external
+        onlyOwner
+    {
+        setMapValue(SEAT_RATE_MAP, _seatNumber, _seatRate);
+    }
+
+    function bookSeat(
+        uint256 _date,
+        uint256 _uniqueId,
+        uint256 _maxAmountToPay
+    ) external {
+        require(address(cbc) == msg.sender, "Must be crosschain call");
+
+        // Check that the calling contract was the travel agency linked to this one from
+        // the source blockchain.
+        // TODO check that the root blockchain is trusted
+        (
+            ,
+            uint256 sourceBlockchainId,
+            address sourceContract
+        ) = decodeAtomicAuthParams();
+        require(
+            sourceContract == approvedTravelAgencies[sourceBlockchainId],
+            "Sender is not an approved travel agency"
+        );
+
+        require(_date != 0, "Date can not be zero");
+        for (uint256 i = 0; i < numSeats; i++) {
+            uint256 rate = getMapValue(SEAT_RATE_MAP, i);
+            // If amount is OK and the room is available.
+            if (
+                !isDoubleMapValueLocked(SEAT_BOOKED_BY_2MAP, i, _date) &&
+                rate <= _maxAmountToPay &&
+                address(
+                    uint160(getDoubleMapValue(SEAT_BOOKED_BY_2MAP, i, _date))
+                ) ==
+                address(0)
+            ) {
+                // Book seat
+                setDoubleMapValue(
+                    SEAT_BOOKED_BY_2MAP,
+                    i,
+                    _date,
+                    uint160(tx.origin)
+                );
+                setMapValue(BOOKING_REF_TO_SEAT_NUMBER, _uniqueId, i);
+                setMapValue(BOOKING_REF_TO_DATE, _uniqueId, _date);
+                // Pay for room.
+                erc20.transferFrom(
+                    travelAgencySpender[sourceContract],
+                    owner,
+                    rate
+                );
+                return;
+            }
+        }
+        require(false, "No seats available");
+    }
+
+    function getBookingInformation(uint256 _uniqueId)
+        external
+        view
+        returns (
+            uint256 amountPaid,
+            uint256 seatId,
+            uint256 date
+        )
+    {
+        date = getMapValue(BOOKING_REF_TO_DATE, _uniqueId);
+        if (date == 0) {
+            amountPaid = 0;
+            seatId = 0;
+        } else {
+            seatId = getMapValue(BOOKING_REF_TO_SEAT_NUMBER, _uniqueId);
+            amountPaid = getMapValue(SEAT_RATE_MAP, seatId);
```

Comment:
could the seat rate change after a booking is made `changeSeatRate()`? In that case would the `amountPaid` not represent what was actually paid for the booking?